File: 0802-find-eventual-safe-states/0802-find-eventual-safe-states.cpp
class Solution {
public:
    vector<int> eventualSafeNodes(vector<vector<int>>& graph) {
        vector <int> ans;
        queue <int> q;
        unordered_map <int, vector <int>> rev;
        vector <int> out(graph.size(), 0);
        for (int i = 0; i < graph.size(); i++) {
            if (graph[i].empty()) {
                q.push(i);
            }
            for (auto nei : graph[i]) {
                rev[nei].push_back(i);
            }
            out[i] = graph[i].size();
        }
        while (!q.empty()) {
            int node = q.front();
            q.pop();
            ans.push_back(node);
            for (auto nei : rev[node]) {
                out[nei]--;
                if (out[nei] == 0) {
                    q.push(nei);
                }
            }
        }
        sort(ans.begin(), ans.end());
        return ans;
    }
};

File: 0802-find-eventual-safe-states/0802-find-eventual-safe-states.go
func eventualSafeNodes(g [][]int) []int {
    s := make([]int, len(g))

    var dfs func(int) bool
    dfs = func(node int) bool {
        if s[node] > 0 {
            return s[node] == 2
        }
        s[node] = 1
        for _, nei := range g[node] {
            if !dfs(nei) {
                return false
            }
        }
        s[node] = 2
        return true
    }

    ans := make([]int, 0)

    for i := range len(g) {
        if dfs(i) {
            ans = append(ans, i)
        }
    }

    return ans
}